File: iosApp/iosApp/CountryViewModel.swift
//
//  CountryViewModel.swift
//  iosApp
//
//  Created by SS on 12/02/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

enum CountryState {
    case loading
    case failed
    case loaded(countries: [Country_])
}

class CountryViewModel: ObservableObject {

    @Published
    private(set) var countryState: CountryState = .loading
    
    private let countryUseCase: LoadCountryUseCase
    
    init(countryUseCase: LoadCountryUseCase) {
        self.countryUseCase = countryUseCase
        fetchCountries()
    }
    
    func fetchCountries() {
        countryUseCase.invoke { (result : Result?, error : Error?) in
            if let successResult = result as? ResultSuccess<AnyObject> {
                let countries = successResult.data as? [Country_]
                self.countryState = .loaded(countries: countries ?? [])
            } else {
                self.countryState = .failed
            }
            
            if result is ResultError<AnyObject> {
                self.countryState = .failed
            }
        }
    }
    
}

File: androidApp/src/main/java/com/ssverma/theearth/android/CountryViewModel.kt
package com.ssverma.theearth.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssverma.theearth.domain.Result
import com.ssverma.theearth.domain.model.Country
import com.ssverma.theearth.domain.usecase.LoadCountryUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

sealed interface CountryState {
    object Loading : CountryState

    data class Loaded(
        val countries: List<Country>
    ) : CountryState

    object Failed : CountryState
}

class CountryViewModel constructor(
    private val countryUseCase: LoadCountryUseCase = LoadCountryUseCase()
) : ViewModel() {

    private val _countryState = MutableStateFlow<CountryState>(CountryState.Loading)
    val countryState = _countryState.asStateFlow()

    init {
        viewModelScope.launch {
            when (val countryResult = countryUseCase()) {
                is Result.Error -> {
                    _countryState.update { CountryState.Failed }
                }
                is Result.Success -> {
                    _countryState.update { CountryState.Loaded(countries = countryResult.data) }
                }
            }
        }
    }
}